Internal: Nested elements. [ED-2256] (#18030)
elementor/elementor: docs/modules/tabs-v2/index.md

  > __TabsV2__ is a module that allows you to created nested tabs.

  * __TabsV2__ module includes:
    * Editor scripts:
        * Widget: __TabsV2__ 
        * Custom Views: 
            * __View__ `assets/js/editor/views/view.js` - The actual view of the widget. the code here is determine the clicks inside the widget Why how?
        * Custom Empty widget views:
            * __Empty View__ `assets/js/editor/views/empty.js` - The view that will be rendered when the widget is empty.
            * __Select Preset View__ `assets/js/editor/views/select-preset.js` - will be rendered when select preset is selected.
            * __Add Section Area View__ `assets/js/editor/views/add-section-area.js` - The default that will be renderd on the __Empty View__.
    * Frontend scripts:
        * __TabsV2__ Handler `assets/js/frontend/handlers/tabs-v2.js`
    * Frontend styles:
        * __TabsV2__ Styles `assets/scss/frontend.scss`
    * Backend Widget:
        * __TabsV2__ Widget registration `widgets/tabs-v2.php`

Let start by registering the module:

## `- Module.php` - How to register a module.
* **Link to the actual file** - [module.php](../../../modules/tabs-v2/module.php)

> * **Description** - The module, enable the experiment on/off, register editor scripts

Reviewer: ```suggestion
* **Description** - Registers the experiment & editor scripts
```

* **Extends** - `\Elementor\Core\Base\Module`

How to register a module?

> * Since __TabsV2__ (nested tabs) is depend on `NestedElements` module, 

Reviewer: ```suggestion
* Since __TabsV2__ (nested tabs) depends on `NestedElements` module, 
```

>    - use `get_experimental_data` function is used to notify the module dependency upon `NestedElementsModule`. please see `'dependencies'` key.

Reviewer: ```suggestion
   - `get_experimental_data` method is used to notify the module dependency upon `NestedElementsModule`. Please see `'dependencies'` key.
```

    ```php
    use Elementor\Modules\NestedElements\Module as NestedElementsModule;
  
    public static function get_experimental_data() {
        return [
            'name' => 'tabs-v2',
            'title' => esc_html__( 'Nested Tab', 'elementor' ),
            'description' => esc_html__( 'Nested Tabs', 'elementor' ),
            'release_status' => Experiments_Manager::RELEASE_STATUS_ALPHA,
            'default' => Experiments_Manager::STATE_INACTIVE,

>             'dependencies' => [ NestedElementsModule::class ], // Tells the expeirmient is dependent upon NestedElementsModule

Reviewer: ```suggestion
            'dependencies' => [ NestedElementsModule::class ],
```

        ];
    }
    ```
* Loading the editor scripts.
  ```php
  public function __construct() {
        add_action( 'elementor/editor/before_enqueue_scripts', function () {
            // The script you load for the editor goes here.
            wp_enqueue_script( 'tabs-v2', $this->get_js_assets_url( 'tabs-v2' ), [
                'elementor-common',
            ], ELEMENTOR_VERSION, true );
        } );
  }
    ```
  * Does it is a requirement for creating nested widgets?
    * Yes, it is, since nested elements are require different `Model` to allow child to be another widget. 

* To register the widget, extend the `get_widgets` method in the `Module` and return the widget name, eg:
  ```php

>   protected function get_widgets() {
>             return [ 'TabsV2' ]; // Located at widgets/tabs-v2.php (the file will be loaded automatically).
>   }

Reviewer: ```suggestion
  protected function get_widgets() {
      return [ 'TabsV2' ]; // Located at widgets/tabs-v2.php (the file will be loaded automatically).
  }
```

                  ],
              ],
              [
                  'elType' => 'container',
                  'settings' => [
                      '_title' => __( 'Tab #2', 'elementor' ),
                  ],
              ],
          ];
      }
	
      protected function get_default_repeater_title_setting_key() {
          return 'tab_title';
      }
	
      protected function get_defaults_children_title() {
          return esc_html__( 'Tab #%d', 'elementor' );
      }
	
      protected function get_default_children_placeholder_selector() {
          return '.elementor-tabs-content-wrapper';
      }
	
      protected function get_html_wrapper_class() {
          return 'elementor-widget-tabs-v2';
      }
  }
  ```
  
## `assets/js/editor/index.js` - Load the module.
* **Link to the actual file** - [index.js](../../../modules/tabs-v2/assets/js/editor/index.js)

* This is first loaded file in the editor.
  * What it does? Await for `Nested Elements` module to be loaded (requirement since the TabsV2 depends on the Nested Elements module).
  * Load the __TabsV2__ module.
      ```javascript
      // On editor init components.

>       elementorCommon.elements.$window.on( 'elementor:init-components', async () => {

Reviewer: ```suggestion
      elementorCommon.elements.$window.on( 'elementor/init-components', async () => {
```

	* The new nested infrastructure is allowing to avoid it, if you know the markup in advance using new feature called __partial render__:
		* The feature will avoid the full re-render and select this exact node/markup and modify only it.
		* You will be able to achieve that using the following markup in the element:
			- For repeater items, example:
      ```html
        <element 
            'data-binding-type': 'repeater-item',  // Type of binding (to know how to behave).
            'data-binding-repeater-name': 'tabs',  // Repeater setting key that effect the binding.
            'data-binding-setting': 'tab_title',   // The key in the repeater that effect the binding.
            'data-binding-index': tabCount,        // Index is required for repeater items.
        >
        </element>
      ```
        - For simple settings items, example:
	  ```html
		 <element
		 	'data-binding-type': 'content', 		 // Type of binding (to know how to behave).
		 	'data-binding-setting': 'testimonial_content',   // Setting change to capture, the value will replace the data-binding.
		 </element>
	  ```

>       > Use it in the `_content_template()` method.

Reviewer 2: Add an explanation about the global defaults in CSS variables

Author: Handled.
